soc: espressif: esp32c6: Add LP_CORE support

LucasTambor opened the pull request:
* Add ULP Coprocessor support for C6.
This requires a change in the board qualifier depending on the build
target.
* Update esp32c6 overlay and configuration files to the proper name.
* Add ULP directory for samples related to the Ultra Low Power Coprocessor.
* Add a sample to demonstrate the debug process of the LP Core

LucasTambor committed:
samples: boards: espressif: ulp

Add ULP directory for samples related to the
Ultra Low Power Coprocessor.
Add a sample to demonstrate the debug process of the LP Core

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

martinjaeger commented on include/zephyr/dt-bindings/clock/esp32c6_clock.h:
```diff
@@ -44,44 +44,49 @@
-#define ESP32_BT_MODULE               33
-#define ESP32_IEEE802154_MODULE       34
-#define ESP32_COEX_MODULE             35
-#define ESP32_PHY_MODULE              36
-#define ESP32_MODULE_MAX              37
+#define ESP32_WIFI_MODULE                33
+#define ESP32_BT_MODULE                  34
+#define ESP32_IEEE802154_MODULE          35
+#define ESP32_COEX_MODULE                36
+#define ESP32_PHY_MODULE                 37
```
How come that the values are suddenly changing? Was this a bug here or in ESP-IDF? I'm asking because we are currently struggling with weird interrupt and clock setup issues for the IEEE 802.15.4 driver and this may be related.

LucasTambor replied:
The values here can change according to HAL version, depending on new peripheral support. The value is mapped in this function: https://github.com/zephyrproject-rtos/hal_espressif/blob/zephyr/components/hal/esp32c6/include/hal/clk_gate_ll.h#L21

dcpleung commented:
Please split the commits: one for SoC changes, one for board, and one for each test.

LucasTambor commented:
> Please split the commits: one for SoC changes, one for board, and one for each test.

@dcpleung Wouldn't that affect bisectability? This PR doesn’t just add a new target; it also modifies an existing one.

dcpleung commented:
> > Please split the commits: one for SoC changes, one for board, and one for each test.
> 
> @dcpleung Wouldn't that affect bisectability? This PR doesn’t just add a new target; it also modifies an existing one.

If you extract the renaming of hpcore into another commit, the lpcore introduction can then be divided into smaller pieces.

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,128 @@
+		help
+		  Choose the ULP Coprocessor type: ULP FSM (Finite State Machine) or ULP RISC-V.
+
+	config ULP_COPROC_TYPE_FSM
+		bool "ULP FSM (Finite State Machine)"
+		depends on SOC_SERIES_ESP32 || SOC_SERIES_ESP32S2 || SOC_SERIES_ESP32S3
+	config ULP_COPROC_TYPE_RISCV
+		bool "ULP RISC-V"
+		depends on SOC_SERIES_ESP32S2 || SOC_SERIES_ESP32S3
+	config ULP_COPROC_TYPE_LP_CORE
+		bool "LP core RISC-V"
+		depends on SOC_SERIES_ESP32C6
+	endchoice
```
newline gaps needed

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,128 @@
+		depends on SOC_SERIES_ESP32C6
+	endchoice
+
+	config ULP_COPROC_RESERVE_MEM
+		int
+		prompt "RTC slow memory reserved for coprocessor"
+		depends on ULP_COPROC_ENABLED
+		default $(dt_node_reg_size_int,$(DT_LPCORE_PARTITION))
+		range 32 8176 if SOC_SERIES_ESP32 || SOC_SERIES_ESP32S2 || SOC_SERIES_ESP32S3
+		range 32 16352 if SOC_SERIES_ESP32C6
+		help
+		  Bytes of memory to reserve for ULP Co-processor firmware & data.
+		  Data is reserved at the beginning of RTC slow memory.
```
not sure why this is Kconfig, it's dts and comes from dts, just use the dts value directly?

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,128 @@
+		help
+		  Turn on this setting to enabled interrupts on the ULP RISC-V core.
+
+	config ULP_RISCV_UART_BAUDRATE
+		int
+		prompt "Baudrate used by the bitbanged ULP RISC-V UART driver"
+		default 9600
+		help
+		  The accuracy of the bitbanged UART driver is limited, it is not
+		  recommend to increase the value above 19200.
```
dts

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,128 @@
+	config ULP_RISCV_I2C_RW_TIMEOUT
+		int
+		prompt "Set timeout for ULP RISC-V I2C transaction timeout in ticks."
+		default 500
```
this should be in the driver Kconfig, and depend on the driver being enabled

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,128 @@
+		  is invoked on.
+endmenu
+
+	config ULP_SHARED_MEM
+		depends on ULP_COPROC_TYPE_LP_CORE
+		hex
+		default 0x10
+		help
+		  Size of the shared memory defined in ulp_lp_core_memory_shared.c.
+		  Size should be kept in-sync with the size of the struct defined there.
```
dts

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,128 @@
+
+	config ULP_HP_UART_CONSOLE_PRINT
+		depends on ULP_COPROC_TYPE_LP_CORE
+		bool
```
types go first, help goes last, rest go in-between

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,128 @@
+menu "ULP Debugging Options"
+	config ULP_PANIC_OUTPUT_ENABLE
+		depends on ULP_COPROC_TYPE_LP_CORE
+		bool
+		prompt "Panic handler outputs to LP UART"
```
there is no need for prompt on these, combine it with the type line e.g. `bool "Panic handler outputs to LP UART"`

nordicjm commented on soc/espressif/esp32c6/CMakeLists.txt:
```diff
@@ -42,18 +51,26 @@ if(NOT CONFIG_BOOTLOADER_MCUBOOT)
 endif()
 
+if(CONFIG_SOC_ESP32C6_HPCORE)
 # get code-partition slot0 address
 dt_nodelabel(dts_partition_path NODELABEL "slot0_partition")
+elseif(CONFIG_SOC_ESP32C6_LPCORE)
+dt_nodelabel(dts_partition_path NODELABEL "lpcore_partition")
```
2 space indent

nordicjm commented on soc/espressif/esp32c6/CMakeLists.txt:
```diff
@@ -42,18 +51,26 @@ if(NOT CONFIG_BOOTLOADER_MCUBOOT)
+    board_finalize_runner_args(esp32 "--esp-app-address=${boot_off}")
+  endif()
+elseif(CONFIG_SOC_ESP32C6_LPCORE)
+    board_finalize_runner_args(esp32 "--esp-app-address=${img_0_off}")
```
2 not 4

nordicjm commented on soc/espressif/esp32c6/CMakeLists.txt:
```diff
@@ -71,6 +88,17 @@ endif()
+  set(SOC_LINKER_SCRIPT ${CMAKE_CURRENT_SOURCE_DIR}/default_lpcore.ld CACHE INTERNAL "")
+  set(CMAKE_C_FLAGS "-Os -march=rv32imac__zicsr_zifencei -mdiv -fdata-sections -ffunction-sections"
+    CACHE STRING "C Compiler Base Flags")
+set(CMAKE_CXX_FLAGS "-Os -march=rv32imac__zicsr_zifencei -mdiv -fdata-sections -ffunction-sections"
```
wonky indent and don't see why these flags should even be here @tejlmand

LucasTambor replied:
This is some test leftovers, thanks for spotting them!

nordicjm reviewed (changes requested):
this is a huge commit touching multiple areas

sylvioalves commented on samples/boards/espressif/ulp/lp_core/debug_ulp/sysbuild.cmake:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright 2024 Espressif
```
2025

raffarost commented on boards/espressif/esp32c6_devkitc/esp32c6_devkitc_lpcore.dts:
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2023 Espressif Systems (Shanghai) Co., Ltd.
```
maybe 2025 :)

raffarost commented on dts/riscv/espressif/esp32c6/esp32c6_lpcore_wroom_n4.dtsi:
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2023 Espressif Systems (Shanghai) Co., Ltd.
```
2025 since it's a new file

LucasTambor committed:
soc: espressif: esp32c6: Add LP Core

Add ULP Coprocessor support for ESP32C6.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
boards: espressif:  esp32c6: Add LP Core board support

Add ULP Coprocessor board support for C6.
This requires a change in the board qualifier depending on the build
target.
Update esp32c6 overlay and configuration files to the proper name.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

raffarost commented:
LGTM, will wait for HAL side update

nordicjm commented on dts/riscv/espressif/esp32c6/esp32c6_common.dtsi:
```diff
@@ -75,9 +75,14 @@
 			compatible = "zephyr,memory-region", "mmio-sram";
 			reg = <0x50000000 DT_SIZE_K(16)>;
 			zephyr,memory-region = "SRAMLP ";
+			shmlp: memory@0 {
```
inconsistent spacing

LucasTambor replied:
why?

nordicjm replied:
No newline before the child node here but there is in the rest of the file

nordicjm commented on dts/riscv/espressif/esp32c6/esp32c6_lpcore.dtsi:
```diff
@@ -0,0 +1,63 @@
+#include <dt-bindings/pinctrl/esp32c6-pinctrl.h>
+
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+
```
why is this needed?

LucasTambor replied:
child node

gmarull replied:
why do all root child nodes need address/size? This doesn't seem to make any sense. You already have `soc` node (`simple-bus`) properly set.

LucasTambor replied:
Sorry! I mistook the lines from the comment.

LucasTambor replied:
@nordicjm This seems to be quite common (https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/common/nordic/nrf54h20.dtsi#L21) and lots of others `.dts` and `.dtsi` . I am not sure why this is needed, but it does trigger a build warning when removing it.

nordicjm commented on soc/espressif/Kconfig:
```diff
@@ -56,4 +56,5 @@ config RTC_CLK_CAL_CYCLES
 endmenu
 
+rsource "Kconfig.ulp"
 endif # SOC_FAMILY_ESPRESSIF_ESP32
```
newline

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,83 @@
+# SPDX-License-Identifier: Apache-2.0
+
+menu "Ultra Low Power (ULP) Co-processor"
+config ULP_COPROC_ENABLED
```
newline

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,83 @@
+endchoice
+
+menu "ULP RISC-V Settings"
+depends on ULP_COPROC_TYPE_RISCV
```
indent

nordicjm commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,83 @@
+endmenu
+
+menu "ULP Debugging Options"
+config ULP_PANIC_OUTPUT_ENABLE
```
newline

nordicjm commented on soc/espressif/esp32c6/hpcore_init_ulp.c:
```diff
@@ -0,0 +1,49 @@
+}
+
+#if defined(CONFIG_ULP_COPROC_ENABLED)
+SYS_INIT(lp_core_image_init, POST_KERNEL, 50);
```
soc hook not sys init

nordicjm commented on soc/espressif/esp32c6/soc_lpcore.c:
```diff
@@ -0,0 +1,42 @@
+	if (sleep_duration_ticks) {
+		ulp_lp_core_lp_timer_set_wakeup_ticks(sleep_duration_ticks);
+	}
+#endif /*SOC_LP_TIMER_SUPPORTED*/
```
```suggestion
#endif /* SOC_LP_TIMER_SUPPORTED */
```
though bit off to have a comment on this endif and not the one directly above it

nordicjm commented on soc/espressif/esp32c6/vectors_lpcore.S:
```diff
@@ -0,0 +1,142 @@
+	csrr  t0, mtval
+	sw    t0, RV_STK_MTVAL(sp)
+
+	csrr	a1, mcause		/* exception cause */
```
Capitalise the first letter of comments, apply throughout

nordicjm commented on boards/espressif/esp32c6_devkitc/esp32c6_devkitc_lpcore_defconfig:
```diff
@@ -0,0 +1,30 @@
+# Copyright (c) 2025 Espressif Systems (Shanghai) Co., Ltd.
+# SPDX-License-Identifier: Apache-2.0
+
+# Single-threaded, no timer support in the kernel
+CONFIG_MULTITHREADING=n
+CONFIG_NUM_PREEMPT_PRIORITIES=0
+CONFIG_SYS_CLOCK_EXISTS=n
+
+# Dynamic memory
+CONFIG_KERNEL_MEM_POOL=n
+CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=0
+
+# Interrupts
+CONFIG_DYNAMIC_INTERRUPTS=n
+CONFIG_IRQ_OFFLOAD=n
+
+# Memory protection
+CONFIG_THREAD_STACK_INFO=n
+CONFIG_THREAD_CUSTOM_DATA=n
+
+# Boot
+CONFIG_BOOT_BANNER=n
+
+# Console
+CONFIG_PRINTK=n
+
+# Build
+CONFIG_SIZE_OPTIMIZATIONS=y
+
+CONFIG_GEN_ISR_TABLES=n
```
some of these seem to be soc specific so should be set in Kconfig.defconfig at soc level. For malloc size, does having that enabled cause the build to fail to run/why is it being set to 0? Likewise with boot banner, printk, etc. why and these being forced off?

LucasTambor replied:
In LP Core we have only 16KB of memory. We set these configurations off to reduce the app size. We set `CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE` to 0 because we want to avoid `malloc_prepare` at the moment. Heap mgmt will come in later PRs. Printk, banner, etc are not set because there are no UART support, these and other peripherals support will come in later PRs also.

nordicjm commented on soc/espressif/esp32c6/CMakeLists.txt:
```diff
@@ -42,18 +51,26 @@ if(NOT CONFIG_BOOTLOADER_MCUBOOT)
-# get code-partition slot0 address
-dt_nodelabel(dts_partition_path NODELABEL "slot0_partition")
+if(CONFIG_SOC_ESP32C6_HPCORE)
+  # get code-partition slot0 address
```
as per other comment

LucasTambor commented:
There is a test failing related to #84254

nordicjm reviewed (changes requested):
Comments not addressed

nordicjm commented:
Other comments need addressing (and rebase)

nordicjm commented on dts/riscv/espressif/esp32c6/esp32c6_lpcore.dtsi:
```diff
@@ -0,0 +1,64 @@
+		flash: flash-controller@60002000 {
+			compatible = "espressif,esp32-flash-controller";
+			reg = <0x60002000 0x1000>;
+
```
```suggestion
```

nordicjm reviewed (commented):
Change OK, nit.

sylvioalves commented on dts/common/espressif/partitions_0x0_default_4M.dtsi:
```diff
@@ -49,6 +49,11 @@
 			reg = <0x3FF000 DT_SIZE_K(4)>;
 		};
+
+		lpcore_partition: partition@3F1000 {
```
Would you mind make this lower case: `partition@3f1000`?

sylvioalves commented on samples/cpp/hello_world/sample.yaml:
```diff
@@ -22,3 +22,4 @@ tests:
       - intel_ish_5_4_1
       - intel_ish_5_6_0
       - intel_ish_5_8_0
+      - esp32c6_devkitc/esp32c6/lpcore
```
Is it still needed?

LucasTambor replied:
Yep, I couldn't find a better way to filter off this sample

sylvioalves commented on soc/espressif/Kconfig.ulp:
```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2025 Espressif Systems (Shanghai) Co., Ltd.
+# SPDX-License-Identifier: Apache-2.0
+
+menu "Ultra Low Power (ULP) Co-processor"
```
Would be good to use either `Coprocessor` or `Co-processor` along the PR/documentation. I've seen both being used.

sylvioalves commented on soc/espressif/common/Kconfig:
```diff
@@ -26,7 +26,7 @@ config ESP_SIMPLE_BOOT
 
 config ESP_HEAP_RUNTIME
```
This will go after #84331

sylvioalves commented on soc/espressif/esp32c6/soc.c:
```diff
@@ -29,6 +29,9 @@
 #include <zephyr/toolchain/gcc.h>
 #include <soc.h>
 
+#ifndef CONFIG_MCUBOOT
```
Is it being used anywhere? Not in this file I guess?

sylvioalves commented on soc/espressif/esp32c6/soc_lpcore.c:
```diff
@@ -0,0 +1,42 @@
+	ulp_lp_core_update_wakeup_cause();
+
+	/* Start Zephyr */
+	z_cstart();
```
`z_cstart()` usually does not return. Can you confirm whether the code after it is indeed running?

marekmatej commented on soc/espressif/esp32c6/hpcore_init_ulp.c:
```diff
@@ -0,0 +1,46 @@
+	const uint32_t lpcore_img_size = FIXED_PARTITION_SIZE(lpcore_partition);
+	int ret = 0;
+
+	printf("Getting LPU image at %p, size %d\n", (void *)lpcore_img_off, lpcore_img_size);
```
not sure if using `printf` in this stage is best idea.

marekmatej commented on dts/riscv/espressif/esp32c6/esp32c6_lpcore.dtsi:
```diff
@@ -0,0 +1,63 @@
+			clock-source = <ESP32_RTC_FAST_CLK_SRC_XTAL_D2>;
+			clock-frequency = <DT_FREQ_M(20)>;
+			xtal-freq = <DT_FREQ_M(40)>;
+		};
```
Why not adding lpcore cpu under the `cpus` node?

LucasTambor replied:
What do you mean? `cpu0` is lpcore in this context

marekmatej replied:
you're right. I was tricked by `cpu0:spu@0`, shouldn't be this numbered otherwise?

LucasTambor commented on boards/espressif/esp32c6_devkitc/esp32c6_devkitc_lpcore.dts:
```diff
@@ -0,0 +1,19 @@
+/dts-v1/;
+
+#include <espressif/esp32c6/esp32c6_lpcore_wroom_n4.dtsi>
+#include <espressif/partitions_0x0_default.dtsi>
```
`slot0_lpcore_partition` is also defined in `partitions_0x0_amp_4M.dtsi`

sylvioalves commented on soc/espressif/esp32c6/default_lpcore.ld:
```diff
@@ -0,0 +1,133 @@
+        KEEP(*(.shared_mem))
+    } > shared_mem_ram
+
+
```
Can you remove this extra space?

sylvioalves commented on soc/espressif/esp32c6/default_lpcore.ld:
```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2024 Espressif Systems (Shanghai) Co., Ltd.
```
2025 instead?

sylvioalves commented on boards/espressif/esp32c6_devkitc/esp32c6_devkitc_lpcore.yaml:
```diff
@@ -0,0 +1,16 @@
+identifier: esp32c6_devkitc/esp32c6/lpcore
+name: ESP32-C6 Devkitc LPCORE
```
Would you re-write to follow the board's "official" name: `ESP32-C6-DevKitC`

sylvioalves commented on boards/espressif/esp32c6_devkitc/esp32c6_devkitc_hpcore.yaml:
```diff
@@ -1,4 +1,4 @@
-identifier: esp32c6_devkitc
+identifier: esp32c6_devkitc/esp32c6/hpcore
 name: ESP32-C6
```
ESP32-C6-DevKitC

marekmatej commented on west.yml:
```diff
@@ -162,7 +162,7 @@ manifest:
       groups:
         - hal
     - name: hal_espressif
-      revision: fbbe8f22f34cadc43b69f9eb3fca10e301874e36
```
as a good practice, I suggest moving manifest changes as a first commit or part of the first commit.

LucasTambor replied:
That would mess with bissectability, changes in HAL are valid from this commit forward

sylvioalves commented on soc/espressif/esp32c6/Kconfig:
```diff
@@ -4,14 +4,15 @@
 	select RISCV_ISA_EXT_M
 	select RISCV_ISA_EXT_C
 	select RISCV_ISA_EXT_ZICSR
+	select RISCV_ISA_EXT_ZIFENCEI if SOC_ESP32C6_LPCORE
```
ESP32C6 have zifencei extension.. Why filtering it out only to SOC_ESP32C6_LPCORE?

LucasTambor commented:
Ci fix in #87227

kartben commented on samples/boards/espressif/ulp/lp_core/debug_ulp/README.rst:
```diff
@@ -0,0 +1,124 @@
+.. zephyr:code-sample:: debug-ulp
+   :name:
```
Forgot the name :) prob a bug for me to fix that this didn't error out

Need to fix since this makes it quite hard to open the sample! https://builds.zephyrproject.io/zephyr/pr/84092/docs/samples/boards/espressif/ulp/lp_core/README.html

kartben commented on samples/boards/espressif/ulp/lp_core/debug_ulp/README.rst:
```diff
@@ -0,0 +1,124 @@
+.. zephyr:code-sample:: debug-ulp
+   :name:
+   :relevant-api: ipm_interface
```
Are you sure,? Fwiw it's ok to remove the line if needed

kartben commented on boards/espressif/esp32c6_devkitc/doc/index.rst:
```diff
@@ -147,7 +147,7 @@ To build the sample application using sysbuild use the command:
    :tool: west
    :zephyr-app: samples/hello_world
-   :board: esp32c6_devkitc
+   :board: esp32c6_devkitc/esp32c6/hpcore
```
New target names will break existing users. Would recommend adding an entry to migration guide, and you should also add a deprecated board name for `esp32c6_devkit` to boards/deprecated.cmake aliasing the new hpcore target

kartben commented on boards/espressif/esp32c6_devkitc/doc/index.rst:
```diff
@@ -237,17 +237,40 @@ Here is an example for building the :zephyr:code-sample:`hello_world` applicatio
+***********************
+
+The ESP32-C6 SoC has two RISC-V cores: the High-Performance Core (HP CORE) and the Low-Power Core (LP CORE).
+It features ultra low power consumption, an interrupt controller, a debug module and a system bus
```
```suggestion
The LP Core features ultra low power consumption, an interrupt controller, a debug module and a system bus
```

kartben commented on samples/boards/espressif/ulp/lp_core/debug_ulp/README.rst:
```diff
@@ -0,0 +1,123 @@
+.. zephyr:code-sample:: debug-ulp
+   :name: Debug ULP
+
+   Sample to demonstrate how to debug the LP Core in ESP32C6.
```
sorry for not catching earlier, as per sample guidelines "It is recommended to word this as if you were completing
   the sentence "This code sample shows how to ..."" so:

```suggestion
   Debug the LP Core on ESP32C6.
```

virtual +1 otherwise :) this is very cool stuff!

sylvioalves commented on soc/espressif/esp32c6/mcuboot.ld:
```diff
@@ -51,6 +51,12 @@ SECTIONS
     *libzephyr.a:esp_rom_spiflash.*(.literal .text .literal.* .text.*)
     *libkernel.a:device.*(.literal .text .literal.* .text.*)
+
+    *libzephyr.a:heap.*(.literal .text .literal.* .text.*)
+
+    *libkernel.a:kheap.*(.literal .text .literal.* .text.*)
+    *libkernel.a:mempool.*(.literal .text .literal.* .text.*)
+
```
Can you re-check whether this entries here are really necessary?

sylvioalves commented on soc/espressif/esp32c6/vector_table_lpcore.S:
```diff
@@ -0,0 +1,22 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
```
2025

sylvioalves commented on soc/espressif/esp32c6/start_lpcore.S:
```diff
@@ -0,0 +1,25 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
```
2025

sylvioalves commented on samples/boards/espressif/ulp/README.rst:
```diff
@@ -0,0 +1,5 @@
+.. zephyr:code-sample-category:: ulp
```
I think this top-level README.rst can go.

sylvioalves commented on samples/boards/espressif/ulp/lp_core/README.rst:
```diff
@@ -0,0 +1,5 @@
+.. zephyr:code-sample-category:: lp-core
```
Same for this file.

sylvioalves commented on soc/espressif/esp32c6/mcuboot.ld:
```diff
@@ -51,6 +51,7 @@ SECTIONS
     *libzephyr.a:flash_map.*(.literal .text .literal.* .text.*)
     *libzephyr.a:esp_rom_spiflash.*(.literal .text .literal.* .text.*)
     *libkernel.a:device.*(.literal .text .literal.* .text.*)
+
```
Remove extra line.
